app/models/grda_warehouse/auth_policies/source_client_policy.rb

```diff
@@ -0,0 +1,75 @@
+###
+# Copyright 2016 - 2024 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+class GrdaWarehouse::AuthPolicies::SourceClientPolicy < GrdaWarehouse::AuthPolicies::BasePolicy
+  # expose role permissions. Optionally rename the permission
+  [
+    [:can_view_client_name, :can_view_name?],
+    [:can_view_client_photo, :can_view_photo?],
+    [:can_view_full_dob],
+    [:can_view_full_ssn],
+    [:can_view_hiv_status],
+  ].each do |permission, method_name|
+    method_name ||= :"#{permission}?"
+    define_method(method_name) do
+      role_permissions.include?(permission)
+    end
+  end
+
+  # can the user see the full client dash page and additional details?
+  memoize def can_view?
+    return true if role_permissions.include?(:can_view_clients)
+    return roi_authorized? if role_permissions.include?(:can_view_client_enrollments_with_roi)
+
+    false
+  end
+
+  protected
+
+  def validate_resource!(arg)
+    ensure_arg_type!(arg, GrdaWarehouse::Hud::Client)
+    raise ArgumentError 'Must be a source client' if arg.destination?
+  end
+
+  def client_id
+    resource.id
+  end
+
+  def client
+    resource
+  end
+
+  memoize def role_permissions
+    clients = GrdaWarehouse::Hud::Client.where(id: client_id)
+    project_ids = GrdaWarehouse::Hud::Project.joins(:clients).merge(clients).distinct.pluck(:id)
+
+    results = Set.new
+    project_ids.each do |project_id|
+      results.merge(context.project_role_permissions(project_id))
+    end
+    results.merge(context.direct_client_role_permissions(client_id))
+  end
```

Comment:
Can you add some documentation here, something like:
* `project_ids` are the projects for which the client has an enrollment
* `results` is a set of permissions the user has for either the project or the client which would grant them access to this client within a permission context
(does that capture it at all?)
I think the method name `role_permissions` is still confusing to me since those are not **role-based** permissions, but just permissions you you have within an appropriate context, using Access Controls when appropriate and roles when appropriate.